pkg/lock/simple_redis.go

```diff
@@ -64,45 +51,109 @@ type redisLock struct {
 	mu     sync.Mutex
 	key    string
 	token  string
-	log    *logrus.Entry
-	rng    *rand.Rand
+	// readers is the number of readers when the lock is acquired for reading
+	// or 0 when it is unlocked, or -1 when it is locked for writing.
+	readers int
+	log     *logrus.Entry
+	rng     *rand.Rand
 }
 
-func (rl *redisLock) Lock() error {
-	rl.mu.Lock()
-	defer rl.mu.Unlock()
+func (rl *redisLock) extends(writing bool) (bool, error) {
+	if rl.token == "" {
+		return false, nil
+	}
+	if (writing && rl.readers > 0) || (!writing && rl.readers < 0) {
+		return false, nil
+	}
 
-	if rl.token != "" {
-		// we already have a lock, attempts to extends it
-		ttl := strconv.FormatInt(int64(LockTimeout/time.Millisecond), 10)
-		ok, err := rl.client.Eval(luaRefresh, []string{rl.key}, rl.token, ttl).Result()
-		if err != nil {
-			return err // most probably redis connectivity error
-		}
+	// we already have a lock, attempts to extends it
+	ttl := strconv.FormatInt(int64(LockTimeout/time.Millisecond), 10)
+	ok, err := rl.client.Eval(luaRefresh, []string{rl.key}, rl.token, ttl).Result()
+	if err != nil {
+		return false, err // most probably redis connectivity error
+	}
 
-		if ok == int64(1) {
-			return nil
+	if ok == int64(1) {
+		if !writing {
+			rl.readers++
 		}
-		// this lock is unavailable, fallback to creating it
-		rl.token = ""
+		return true, nil
 	}
 
-	token := utils.RandomStringFast(rl.rng, lockTokenSize)
+	return false, nil
+}
+
+func (rl *redisLock) obtains(writing bool, token string) (bool, error) {
+	// Try to obtain a lock
+	ok, err := rl.client.SetNX(rl.key, token, LockTimeout).Result()
+	if err != nil {
+		return false, err // most probably redis connectivity error
+	}
+	if !ok {
+		return false, nil
+	}
+
+	rl.token = token
+	if writing {
+		rl.readers = -1
+	} else {
+		rl.readers++
+	}
+	return true, nil
+}
+
+func (rl *redisLock) extendsWriting() (bool, error) {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	return rl.extends(true)
+}
+
+func (rl *redisLock) obtainsWriting(token string) (bool, error) {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	return rl.obtains(true, token)
+}
+
+func (rl *redisLock) Lock() error {
+	ok, err := rl.extendsWriting()
```

Comment:
Why don't we retry to extend the writing lock like we do for reading locks?

Reply:
For a reading lock, there can be several goroutines. If the lock is currently busy for writing, all the goroutines for reading will try to extend and then obtain the lock (but it will fail). When the writing goroutine release the lock, the first reading goroutine will obtain the lock, and the next goroutines can extend it.

For writing, we can't have several goroutines with the lock at the same time, so it didn't make sense to retry extending the lock. And, in fact, even the first call to extend can introduce bugs and I have preferred to not allow a goroutine to calls several times `Lock` and then calls `Unlock` several times: it must call `Unlock` before calling again `Lock`. It is also more consistent with the in-memory lock.